semgrep/semgrep/target_manager.py

```diff
@@ -73,6 +84,196 @@ def converted_pipe_targets(targets: Sequence[str]) -> Iterator[Sequence[str]]:
         yield out_targets
 
 
+@attr.s(auto_attribs=True)
+class IgnoreLog:
+    """Keeps track of which paths were ignored for what reason.
+
+    Each attribute is a distinct reason why files could be ignored.
+
+    Some reason can apply once per rule; these are mappings keyed on the rule id.
+    """
+
+    target_manager: "TargetManager"
+
+    semgrepignored: Set[Path] = Factory(set)
+    always_skipped: Set[Path] = Factory(set)
+    cli_includes: Set[Path] = Factory(set)
+    cli_excludes: Set[Path] = Factory(set)
+    size_limit: Set[Path] = Factory(set)
+
+    rule_includes: Dict[str, Set[Path]] = Factory(lambda: defaultdict(set))
+    rule_excludes: Dict[str, Set[Path]] = Factory(lambda: defaultdict(set))
+    rule_size_limit: Dict[str, Set[Path]] = Factory(lambda: defaultdict(set))
+
+    @property
+    def size_limited_paths(self) -> Set[Path]:
+        """
+        All paths that were skipped because they were larger than the max_target_bytes.
+
+        This is needed because sometimes a file is not targeted at first
+        when we initially filter files that are too large,
+        but then a rule includes it.
```

Comment:
Is this actually what happens? Shouldn't a rule include only ever results in a smaller or equal number of files. i.e. given set A, passing A through rule_includes should not results in a set with elements not in A.

Reply:
I made this assumption when seeing that we do a second size filter pass after applying the rule includes. Good question though, let me check.

Reply:
Looks like this is necessary for explicit files